### PR TITLE
feat: vercel function in node.js

### PR DESCRIPTION
**Pull request by hebingchang**
This PR set up a RPC server under the same origin as frontend.

### CODE REVIEW

File: api/_misc.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  port: number
+
+  constructor(host = "182.254.145.254", port = 8080) {
+    this.host = host
+    this.port = port
+  }
+
+  send_message = (message) => new Promise((resolve, reject) => {
+    const client = net.createConnection({ host: this.host, port: this.port }, () => {
+      client.write(JSON.stringify(message))
```
[the remark this answers is not in the document]
Comment (hebingchang) [reply api/_misc.ts:29]: I don't think so. It seems there's no reliable way to read raw request body in vercel function.

File: api/_misc.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    })
+
+    client.on("data", (data) => {
+      // response.status(200).json
+      resolve(JSON.parse(data.toString()))
```
[the remark this answers is not in the document]
Comment (hebingchang) [reply api/_misc.ts]: Fair point. I've removed `JSON.parse`.

File: api/version.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import { VercelRequest, VercelResponse } from "@vercel/node"
+import { allowCors, Client } from "./_misc"
+
+const handler = (request: VercelRequest, response: VercelResponse) => {
+  const client = new Client()
+  response.status(200).json({
+    "name": "「无可奉告」Android 版",
+    "addr": `tcp://${client.host}:${client.port}`,
+    "terms_of_service": "http://wukefenggao.cn/code",
+    "rpc_source_code": "https://github.com/skyzh/make-a-fortune/blob/master/tcp_proxy/__main__.py",
```
Comment (skyzh) [inline api/version.ts:10]: Point this URL to expected location of `api/` folder for browse.

File: yarn.lock [hunk 4]
```diff
@@ -7025,6 +7060,11 @@ type-fest@^0.20.2:
   resolved "https://registry.yarnpkg.com/type-fest/-/type-fest-0.20.2.tgz#1bf207f4b28f91583666cb5fbd327887301cd5f4"
   integrity sha512-Ne+eE4r0/iWnpAxD852z3A+N0Bt5RN//NjJwRd2VFHEmrywxf5vsZlh4R6lixl6B+wz/8d+maTSAkN1FIkI3LQ==
 
+typescript@3.9.3:
+  version "3.9.3"
+  resolved "https://registry.npm.taobao.org/typescript/download/typescript-3.9.3.tgz#d3ac8883a97c26139e42df5e93eeece33d610b8a"
```
[the remark this answers is not in the document]
Comment (hebingchang) [reply yarn.lock]: Done.

File: yarn.lock [hunk 5]
```diff
@@ -7025,6 +7060,11 @@ type-fest@^0.20.2:
   resolved "https://registry.yarnpkg.com/type-fest/-/type-fest-0.20.2.tgz#1bf207f4b28f91583666cb5fbd327887301cd5f4"
   integrity sha512-Ne+eE4r0/iWnpAxD852z3A+N0Bt5RN//NjJwRd2VFHEmrywxf5vsZlh4R6lixl6B+wz/8d+maTSAkN1FIkI3LQ==
 
+typescript@3.9.3:
```
[the remark this answers is not in the document]
Comment (hebingchang) [reply yarn.lock:7063]: `typescript@3.9.3` is a dependency of `@vercel/node@1.9.1`.

File: api/version.ts (hunk 3 above)
Comment (hebingchang) [reply api/version.ts]: Done.

**Comment by hebingchang** [comment]
Sometimes the RPC got incomplete TCP payloads. I'm going to work on that.

Mark the PR as draft.

**Comment by skyzh** [comment]
> Sometimes the RPC got incomplete TCP payloads. I'm going to work on that (tomorrow).
> 
> Mark the PR as draft.

Seems normal if TCP connection is reset. As there is no something like `Content-Length` in RPC protocol, we may just let it fail.

**Comment by hebingchang** [comment]
> > Sometimes the RPC got incomplete TCP payloads. I'm going to work on that (tomorrow).
> > Mark the PR as draft.
> 
> Seems normal if TCP connection is reset. As there is no things like `Content-Length` in RPC protocol, we may just let it fail.

Yes, but considering the server will **actively** close the connection after data transfer complete, we could add a buffer to receive all data before the connection reset.

image

Now I think the PR is ready for review and merge.

**Review by skyzh — APPROVED** [review]
We could do something like `readline`, as done in python `tcp_proxy`. Besides that, this patch looks good.

**Pull request merged**